Announcement blogpost for the 1.16 release (#574)
apache/flink-web: _posts/2022-10-10-1.16-announcement.md

---
layout: post
title:  "Announcing the Release of Apache Flink 1.16"
subtitle: ""
date: 2022-10-10T08:00:00.000Z
categories: news
authors:
- godfreyhe:
  name: "Godfrey He"
  twitter: "godfreyhe"

---

Apache Flink continues to grow at a rapid pace and is one of the most active 
communities in Apache. Flink 1.16 had over 230 contributors enthusiastically participating, 
with 19 FLIPs and 900+ issues completed, bringing a lot of exciting features to the community.

> Flink has become the leading role and factual standard of stream processing, 
> and the concept of the unification of stream (aka unbounded) and batch (aka bounded) data 

Reviewer: ```suggestion
Flink has become the leading role and factual standard of stream processing. The concept of the unification of stream and batch data 
```
I think a @knaufk had a good point here about not mixing up the concepts of bound/unbound data vs processing types, i.e. stream and batch processing, in [his comment in the GoogleDocs draft](https://docs.google.com/document/d/1rIBNpzJulqEKJCuYtWtG-vDmSsGpN9sip_ewpzMequ0/edit?disco=AAAAf9slSt0) WDYT?

Author: make sense

processing is gradually gaining recognition and is being successfully implemented in more 
and more companies. Previously, the integrated stream and batch concept placed more emphasis 
on a unified API and a unified computing framework. This year, based on this, Flink proposed 
the next development direction of [Flink-Streaming Warehouse](https://www.alibabacloud.com/blog/more-than-computing-a-new-era-led-by-the-warehouse-architecture-of-apache-flink_598821) (Streamhouse), 
which further upgraded the scope of stream-batch integration: it truly completes not only 
the unified computation but also unified storage, thus realizing unified real-time analysis.

> 

Reviewer: ```suggestion
  {% toc %}
```

What about using a table of content here? There is a lot of content with quite a few different sections. A table of content might help navigating through the blog post. Of course, that requires coming up with a reasonable structure for the headlines.

Author: I think it's hard to unify them into a table because they describe the content in different dimensions.

Reviewer: Not sure if I understand your concern correctly. :thinking: Table of content would be just a collection of links to the different sections in the blog post. ...something like an overview of all the topics with a link

Author: do you want the following table ?
![image](https://user-images.githubusercontent.com/8777671/194763558-c05a00a1-afe5-44a1-add7-393e878174f1.png)

Reviewer: No, I guess we're talking about different things. By adding the line I suggested above (see my [first comment in this thread](https://github.com/apache/flink-web/pull/574#discussion_r986763363)) an automatic [table of contents](https://en.wikipedia.org/wiki/Table_of_contents) is generated that might help users to navigate through the page. I'm sorry that this misunderstanding made you generate this table from your previous comment now. :-/
![Screenshot from 2022-10-10 15-10-20](https://user-images.githubusercontent.com/1101012/194874249-31234f82-d606-4892-85c6-000b8a4e6cbf.png)

Author: I get it, thank you for the detailed explanation.  I winder whether we need this catalog, because all the other posts don't have it.

Reviewer: Up to you: It was just an idea to manage the large amount of subsections. :+1:

> In 1.16, the Flink community has completed many improvements for both batch and stream:

Reviewer 2: ```suggestion
In 1.16, the Flink community has completed many improvements for both batch and stream processing:
```
nit to make consistent with text bellow

Author: good catch

    - Changelog State Backend provides users with second or even millisecond checkpoints to 
  dramatically improve the fault tolerance experience, while providing a smaller end-to-end 
  latency experience for transactional Sink jobs.
  - Lookup join is widely used in stream processing. Slow lookup speed, low throughput and 
  delay update are resolved through common cache mechanism, asynchronous io and retriable lookup. 
  These features are very useful, solving the pain points that users often complain about, 
  and supporting richer scenarios.
  - From the first day of the birth of Flink SQL, there were some non-deterministic operations 
  that could cause incorrect results or exceptions, which caused great distress to users. 
  In 1.16, we spent a lot of effort to solve most of the problems, and we will continue to 
  improve in the future.

With the further refinement of the integration of stream and batch, and the continuous iteration of 
the Flink Table Store ([0.2 has been released](https://flink.apache.org/news/2022/08/29/release-table-store-0.2.0.html)), 
the Flink community is pushing the Streaming warehouse from concept to reality and maturity step by step.

> ### Understanding Streaming Warehouses

Reviewer: Is there a reason why we're using level 3 and 4 headlines instead of level 1 and 2? The post looks like a huge text block without real visual separators. Using level 1 headlines might help clustering the text from a visual standpoint. WDYT?

Author: make sense. I refer to the post of 1.15, I will change it.

the existing state backends and improving the usability of changelog state backend:
 - Support state migration
 - Support local recovery
 - Introduce file cache to optimize restoring
 - Support switch based on checkpoint
 - Improve the monitoring experience of changelog state backend
   - expose changelog's metrics
   - expose changelog's configuration to webUI

Table 1: The comparison between Changelog Enabled / Changelog Disabled on value state 
(see [this blog](https://flink.apache.org/2022/05/30/changelog-state-backend.html) for more details)

| Percentile  | End to End Duration | Checkpointed Data Size<sup>*<sup> | Full Checkpoint Data Size<sup>*<sup> |
|-------------|---------------------|-----------------------------------|--------------------------------------|
| 50%         | 311ms / 5s          | 14.8MB / 3.05GB                   | 24.2GB / 18.5GB                      |
| 90%         | 664ms / 6s          | 23.5MB / 4.52GB                   | 25.2GB / 19.3GB                      |
| 99%         | 1s / 7s             | 36.6MB / 5.19GB                   | 25.6GB / 19.6GB                      |

> | 99.9%       | 1s / 10s            | 52.8MB / 6.49GB                   | 25.7GB / 19.8GB                      |

Reviewer: The formatting of the table is odd: There are no lines separating the columns. :thinking:

Author: The line are too long that it was put into two lines in this page. It looks like as bellow in IDE
![image](https://user-images.githubusercontent.com/8777671/194456382-63299cc7-3a9d-4151-8a6f-32a02a86d9c0.png)

Reviewer: Sorry for not being clear enough. I was not referring to the markdown code but the actually rendered table. See the picture below. But it looks like a fundamental issue. I checked other blog post. And we always customize the table layout to add borders in someway. The neatest way (used in the [blog post about release the Kubernetes Flink Operator 1.0.0](https://flink.apache.org/news/2022/06/05/release-kubernetes-operator-1.0.0.html)) I found so far was adding the following line:
```
{:class="table table-bordered"}
```
![Screenshot from 2022-10-07 08-23-01](https://user-images.githubusercontent.com/1101012/194481445-1e01af7d-683b-4e28-93a2-998f29691dac.png)

Reviewer: I created FLINK-29533 to fix the issue generally.

## RocksDB rescaling improvement & rescaling benchmark

Rescaling is a frequent operation for cloud services built on Apache Flink, this release leverages 
[deleteRange](https://rocksdb.org/blog/2018/11/21/delete-range.html) to optimize the rescaling of 
Incremental RocksDB state backend. deleteRange is used to avoid massive scan-and-delete operations, 
for upscaling with a large number of states that need to be deleted, 
the speed of restoring can be almost [increased by 2~10 times](https://github.com/apache/flink/pull/19033#issuecomment-1072267658).

<center>
<img src="{{ site.baseurl }}/img/rocksdb_rescaling_benchmark.png" style="width:90%;margin:15px">
</center>

## Improve monitoring experience and usability of state backend

This release also improves the monitoring experience and usability of state backend.
Previously, RocksDB's log was located in its own DB folder, which made debugging RocksDB not so easy. 
This release lets RocksDB's log stay in Flink's log directory by default. RocksDB statistics-based 
metrics are introduced, statistics-based metrics held at the database level, e.g. total block cache hit 
count within the DB.

## Support overdraft buffer

> A new concept of [overdraft network buffer](https://nightlies.apache.org/flink/flink-docs-master/docs/deployment/memory/network_mem_tuning/#overdraft-buffers) 

Reviewer 2: ```suggestion
A new concept of [overdraft network buffer](https://nightlies.apache.org/flink/flink-docs-release-1.16/docs/deployment/memory/network_mem_tuning/#overdraft-buffers) 
```
Some doc links to master, and some to 1.16 release. Should we standardise to 1.16 docs?

Author: good catch

Author: yes

is introduced to mitigate the effects of uninterruptible blocking a subtask thread during back pressure, 
which can be turned on through taskmanager.network.memory.max-overdraft-buffers-per-gate. 
Starting from 1.16.0, Flink subtask can request by default up to 5 extra (overdraft) buffers over 
the regular configured amount.  This change can slightly increase memory consumption of 
the Flink Job but vastly reduce the checkpoint duration of the unaligned checkpoint.
If the subtask is back pressure by downstream subtasks and the subtask requires more than 
a single network buffer to finish what it's currently doing, overdraft buffer comes into play.

> Read more about this in [the documentation](https://nightlies.apache.org/flink/flink-docs-master/docs/deployment/memory/network_mem_tuning/#overdraft-buffers).

Reviewer 2: ```suggestion
Read more about this in [the documentation](https://nightlies.apache.org/flink/flink-docs-release-1.16/docs/deployment/memory/network_mem_tuning/#overdraft-buffers).
```

## Timeout aligned to unaligned checkpoint barrier in the output buffers of an upstream subtask

This release updates the timing of switching from Aligned Checkpoint (AC)  to Unaligned Checkpoint (UC).

> With UC enabled, if [execution.checkpointing.aligned-checkpoint-timeout](https://nightlies.apache.org/flink/flink-docs-master/docs/ops/state/checkpointing_under_backpressure/#aligned-checkpoint-timeout) 

Reviewer 2: ```suggestion
With UC enabled, if [execution.checkpointing.aligned-checkpoint-timeout](https://nightlies.apache.org/flink/flink-docs-release-1.16/docs/ops/state/checkpointing_under_backpressure/#aligned-checkpoint-timeout) 
```

is configured, each checkpoint will still begin as an AC, but when the global checkpoint duration 
exceeds the aligned-checkpoint-timeout, if the AC has not been completed, 
then the checkpoint will be switched to unaligned.

Previously, the switch of one subtask needs to wait for all barriers from upstream. 
If the back pressure is served, the downstream subtask may not receive all barriers 

> within [checkpointing-timeout](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/deployment/config/#execution-checkpointing-timeout), 

Reviewer 2: ```suggestion
within [checkpointing-timeout](https://nightlies.apache.org/flink/flink-docs-release-1.16/docs/deployment/config/#execution-checkpointing-timeout), 
```

causing the checkpoint to fail.

In this release, If the barrier cannot be sent from the output buffer to the downstream task 
within the execution.checkpointing.aligned-checkpoint-timeout, Flink lets upstream subtasks 
switch to UC first to send barriers to downstream, thereby decreasing 
the probability of checkpoint timeout during back pressure.

> More details can be found in [this documentation](https://nightlies.apache.org/flink/flink-docs-master/docs/ops/state/checkpointing_under_backpressure/#aligned-checkpoint-timeout).

Reviewer 2: ```suggestion
More details can be found in [this documentation](https://nightlies.apache.org/flink/flink-docs-release-1.16/docs/ops/state/checkpointing_under_backpressure/#aligned-checkpoint-timeout).
```

## Non-Determinism In Stream Processing

Users often complain about the high cost of understanding stream processing. 
One of the pain point is the non-determinisim in stream processing (usually not intuitive) 
which may cause wrong results or errors, it is long lived since the first day 
that flink sql was available.

For complex streaming jobs, now it's possible to detect and resolve potential correctness issues 
before running. If the problems can’t be resolved completely, a detailed message could prompt users to 
adjust the SQL so as to avoid introducing non-deterministic problems. More details can be found 
in [the documentation](https://nightlies.apache.org/flink/flink-docs-release-1.16/docs/dev/table/concepts/determinism).

## Enhanced Lookup Join

Lookup join is widely used in stream processing, and we have introduced several improvements:
- Adds a unified abstraction for lookup source cache and 
  [related metrics](https://cwiki.apache.org/confluence/display/FLINK/FLIP-221%3A+Abstraction+for+lookup+source+cache+and+metric) 
  to speed up lookup queries
- Introduces the configurable asynchronous mode (ALLOW_UNORDERED) via 

>   [job configuration](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/config/#table-exec-async-lookup-output-mode) 

Reviewer 2: ```suggestion
  [job configuration](https://nightlies.apache.org/flink/flink-docs-release-1.16/docs/dev/table/config/#table-exec-async-lookup-output-mode) 
```
